jax/lax/lax_parallel.py

```diff
@@ -616,8 +620,37 @@ def all_gather(x, axis_name):
    [0 1 2 3]
    [0 1 2 3]
    [0 1 2 3]]
+
+  An example of using axis_index_groups, groups split by even & odd device ids:
+
+  >>> x = np.arange(16).reshape(4, 4)
+  >>> print(x)
+  [[ 0.  1.  2.  3.]
+   [ 4.  5.  6.  7.]
+   [ 8.  9. 10. 11.]
+   [12. 13. 14. 15.]]
+  >>> y = jax.pmap(lambda x: jax.lax.all_gather(
+  ... x, 'i', axis_index_groups=[[0, 2], [1, 3]]))(x)
+  >>> print(y)
+  [[[ 0.  1.  2.  3.]
+    [ 8.  9. 10. 11.]]
+   [[ 4.  5.  6.  7.]
+    [12. 13. 14. 15.]]
+   [[ 0.  1.  2.  3.]
+    [ 8.  9. 10. 11.]]
+   [[ 4.  5.  6.  7.]
+    [12. 13. 14. 15.]]]
   """
-  return _allgather(x, 0, psum(1, axis_name), axis_name)
+
+  index = axis_index(axis_name)
+  if axis_index_groups is not None:
+    indices = np.array(axis_index_groups).flatten()
+    axis_index_to_group_index = indices.argsort() % len(axis_index_groups[0])
```

Comment:
This is clever, perhaps too clever! It is not obvious that it works. What is the intended result for all_gather with groups? If `index_groups=[[2, 1], [3, 0]]` then should the result be exactly the same as when the index groups are sorted? (`index_groups=[[1, 2], [0, 3]]`)? Essentially in both cases we should have `to_group_index=[0, 0, 1, 1]`? 

how would you argue that this code is correct?

Reply:
But the intra-group order matters when gathering, no? As an end-user I'd definitely expect the result for the two cases to be different which is only possible if `to_group_index` preserves the ordering of `axis_index_groups`.

Reply:
The semantics is up to us. It is Ok to gather respecting the intra-group order, as you propose. In that case, I recommend changing slightly the example in the documentation to have the index groups not sorted intra-group.

Reply:
Makes sense, updated the example to illustrate the importance of intra-group ordering.